### PR TITLE
fix container overflow read

### PR DESCRIPTION
**Pull request by steple**
nearestKSearch might reduce the size of nn_dists.

### CODE REVIEW

**Comment by mvieth** [comment]
@steple Hi, the only reason I can think of why `nearestKSearch` might return fewer than `searcher_k` points, is because the point cloud contains fewer than `search_k` points. Was this the case for you when you experienced the container overflow? Do you know which search method was used? (was the point cloud organized or not?)

**Comment by steple** [comment]
I don't have answers to your questions readily available.  I ran into this when upgrading compilers, and the new compiler catches this issue in the act (with sanitizers).  I'll try to look into it.

I went for what I thought is obvious fix, but alternatively the could could also assert and explode, I guess.

**Comment by mvieth** [comment]
Ah okay. Which compiler are you using, and which sanitizer? Could this have been a false positive? With static analyzers for example, I have seen several cases where the analyzer has found an issue, but on closer inspection it was a false positive.
I am fine with switching to `nn_dists.size()`, but I would like to fully understand the problem first. Who knows, there could be a problem hiding in one of the search methods, leading to the resize :smile:
The division should be `nn_dists.size() - 1` though (minus, not plus), and in the cpp file, it should stay `distances[cp]` (`cp`, not `iii`).

**Comment by steple** [comment]
> The division should be nn_dists.size() - 1 though (minus, not plus), and in the cpp file, it should stay distances[cp] (cp, not iii).

Done.

File: filters/include/pcl/filters/impl/statistical_outlier_removal.hpp [hunk 1]
```diff
@@ -93,11 +93,11 @@ pcl::StatisticalOutlierRemoval<PointT>::applyFilterIndices (Indices &indices)
       continue;
     }
 
-    // Calculate the mean distance to its neighbors
+    // Calculate the mean distance to its neighbors.
     double dist_sum = 0.0;
-    for (int k = 1; k < searcher_k; ++k)  // k = 0 is the query point
-      dist_sum += sqrt (nn_dists[k]);
-    distances[iii] = static_cast<float> (dist_sum / mean_k_);
+    for (int k = 1; k < nn_dists.size(); ++k) // k = 0 is the query point
```
Comment (mvieth) [inline filters/include/pcl/filters/impl/statistical_outlier_removal.hpp:98]:
```suggestion
    for (std::size_t k = 1; k < nn_dists.size(); ++k) // k = 0 is the query point
```

File: filters/src/statistical_outlier_removal.cpp [hunk 2]
```diff
@@ -227,11 +227,11 @@ pcl::StatisticalOutlierRemoval<pcl::PCLPointCloud2>::generateStatistics (double&
...
     }
 
-    // Minimum distance (if mean_k_ == 2) or mean distance
-    double dist_sum = 0;
-    for (int j = 1; j < mean_k_; ++j)
-      dist_sum += sqrt (nn_dists[j]);
-    distances[cp] = static_cast<float> (dist_sum / (mean_k_ - 1));
+    // Calculate the mean distance to its neighbors.
+    double dist_sum = 0.0;
+    for (int k = 1; k < nn_dists.size(); ++k) // k = 0 is the query point
```
Comment (mvieth) [inline filters/src/statistical_outlier_removal.cpp:232]:
```suggestion
    for (std::size_t k = 1; k < nn_dists.size(); ++k) // k = 0 is the query point
```

**Review by mvieth — COMMENTED** [review]
Thanks for the additional information! Just two minor changes (to fix sign-compare problems), then I am happy with the pull request.

File: filters/src/statistical_outlier_removal.cpp (hunk 2 above)
Comment (steple) [reply filters/src/statistical_outlier_removal.cpp]: Done.

File: filters/include/pcl/filters/impl/statistical_outlier_removal.hpp (hunk 1 above)
Comment (steple) [reply filters/include/pcl/filters/impl/statistical_outlier_removal.hpp]: Done.

**Review by mvieth — APPROVED** [review]
Thank you!

**Comment by steple** [comment]
Thank you for making this quick and easy.

**Pull request merged**